itest.py

```diff
@@ -0,0 +1,832 @@
+
+        if isinstance(param, PurePath):
+            param = str(param).format(os_name=os_choice)
+
+        arg_list.append(param)
+
+    arg_list.extend(bonus_args)
+
+    failed_tests_list = None
+    if test_args is None:
+        tests_to_run = input('Please input test you wish to run (default: all): ').split()
+
+        if len(tests_to_run) != 0:
+            failed_tests_list = tests_to_run
+            failed_tests_list.insert(0, '--tests')
+    else:
+        failed_tests_list = test_args[test_idx:]
+
+    res = None
+    failed_tests_pattern = re.compile('(?<=List of failed tests:\n\t).*')
+
+    max_duplicate_attempts = 3
+    current_duplicate_attempt = 1
+    attempt = 0
+
+    while current_duplicate_attempt <= max_duplicate_attempts:
+        attempt += 1
+
+        logger.info(f'Running attempt=[{attempt!r}] with current_duplicate_attempt=[{current_duplicate_attempt!r} of max_duplicate_attempts=[{max_duplicate_attempts!r}]]')
+
+        if failed_tests_list:
+            running_arg_list = arg_list + failed_tests_list
+        else:
+            running_arg_list = arg_list
+
+        logger.debug(f"Executing running_arg_list=[{running_arg_list!r}]")
+
+        try:
+            #with subprocess.Popen(running_arg_list, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True) as res:
+            #    for line in res.stdout:
+            #        print(line, end='')
+
+            #res = stream_subprocess_output(running_arg_list, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
+            res = subprocess.run(running_arg_list, capture_output=True, text=True)
+        except KeyboardInterrupt:
+            logger.warning('Interupt recieved. Kill test...')
+            arg_list = []
+            d_c = shutil.which('docker-compose')
+            arg_list.append(d_c)
+            arg_list.append('down')
+            subprocess.run(arg_list, cwd=str(project_directory))
+            c = input('Continue? (y/N)')
+            if c.lower() == 'n' or len(c) == 0:
+                break
+            else:
+                continue
+
+        logger.debug(f'res=[{res!r}] of attempt=[{attempt!r}]')
+
+        if res.returncode == 0:
+            sha = SHA_RE.search(res.stdout).group('sha')
+            logger.info(f'Test passed on attempt=[{attempt!r}], sha=[{sha!r}]')
+            break
+
+        failed_tests = failed_tests_pattern.search(res.stdout)
+
+        if failed_tests:
+            new_failed_tests_list = failed_tests.group().strip().split()
+            new_failed_tests_list.insert(0, '--tests')
+            if new_failed_tests_list == failed_tests_list:
+                current_duplicate_attempt += 1
+            else:
+                current_duplicate_attempt = 1
+            failed_tests_list = new_failed_tests_list
+        else:
+            logger.error(f'Failed to parse failed_tests with failed_tests_pattern=[{failed_tests_pattern!r}]')
+            current_duplicate_attempt += 1
```

Comment:
Note that the failed_tests grep fails on the plugins, as we expect irods/irods style failed tests output currently.